doc/specs/stdlib_selection.md

```diff
@@ -0,0 +1,364 @@
+---
+title: Selection Procedures
+---
+
+# The `stdlib_selection` module
+
+[TOC]
+
+## Overview of selection
+
+Suppose you wish to find the value of the kth-smallest entry in an array, or
+the index of that value. While it could be done by sorting the whole array
+using `sort` or `sort_index` from `stdlib_sorting` and then finding the k-th
+entry, that would require O(N x LOG(N)) time. However selection of a single
+entry can be done in O(N) time, so is much faster for large arrays.  This is
+useful, for example, to quickly find the median of an array, or some other
+percentile.
+
+The Fortran Standard Library therefore provides a module, `stdlib_selection`,
+which implements selection algorithms.
+
+## Overview of the module
+
+The module `stdlib_selection` defines two generic subroutines:
+* `select` is used to find the kth-smallest entry of an array. The input
+array is also modified in-place, and on return will be partially sorted
+such that `all(array(1:k) <= array(k)))`  and `all(array(k) <= array((k+1):size(array)))`.
+The user can optionally specify `left` and `right` indices to constrain the search
+for the kth-smallest value. This can be useful if you have previously called `select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`array`, thus implying some constraint on the location).
+
+* `arg_select` is used to find the index of the kth-smallest entry of an array.
+In this case the input array is not modified, but the user must provide an
+input index array with the same size as `array`, having unique indices from
+`1:size(array)`, which is modified instead. On return the index array is modified
+such that `all(array(index(1:k)) <= array(index(k)))` and `all(array(k) <= array(k+1:size(array)))`.
+The user can optionally specify `left` and `right` indices to constrain the search
+for the kth-smallest value. This can be useful if you have previously called `arg_select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`index`, thus implying some constraint on the location).
+
+#### Licensing
+
+The Fortran Standard Library is distributed under the MIT
+License. However components of the library may be based on code with
+additional licensing restrictions. In particular `select` and `arg_select`
+were derived by modifying a matlab implementation of "qselect" by Manolis
+Lourakis, https://www.mathworks.com/matlabcentral/fileexchange/68947-qselect
+Below is the licence of the matlab qselect
+
+Copyright (c) 2018, Manolis Lourakis
```

Comment:
Question: I am not an expert in licensing. Is this license compatible with the MIT license?

Reply:
I'm also not an expert on this. 

I'll try to email the matlab code author, and seek explicit permission to release the modified version under an MIT licence while still crediting the original implementation in the comments. A similar thing was done for the some of the sorting routines. 

If that's not possible, it won't be too difficult to develop an alternative implementation. The interface and initial checks in the pull-request are all original. It is just the "partition while-loop" that mirrors the matlab code. This is a pretty standard operation that could be replaced with a new implementation (e.g. informed by coretran, or something else).

Reply:
Same, i'm not a lawyer.... Getting the permission from the author would allow a re-release under a different license than the original.  
The alternative might be to simply state that code written explicitly for stdlib is released under the MIT license, but some parts (translations like this) may have their own license.  I like this [image](https://dwheeler.com/essays/floss-license-slide.html) when referring to licenses.  This selection license looks more like a BSD3 license, and would therefore not be compatible with stdlib **If stdlib was released entirely under MIT**

I know there are stipulations with the less permissive GPL licenses that the entire library would have to be released under GPL, but I don't see any such stipulation in the MIT or BSD (Or this selection algo) text.  Does this not mean a library can be piece-by-piece licensed?  Maybe this should move to another thread?

Reply:
I just emailed the original author, to seek permission to release under MIT while citing the original work.

Reply:
> I just emailed the original author, to seek permission to release under MIT while citing the original work. That will likely be easier than figuring out the legal details.

Good. Thank you.
Otherwise, I guess that some parts of stdlib `sort` could be used.

Reply:
I just implemented changes to the code so it draws on the Coretran implementation of @leonfoks rather than the matlab implementation. 

Coretran has a BSD3 licence -- actually this is very similar to the licence of the matlab code, so likely there would not have been problems. But it is good to make sure.